TFNetworkLayer.py

```diff
@@ -2036,7 +2036,7 @@ def get_out_data_from_opts(cls, name, sources=(), extra_deps=(), out_type=None,
     :param int|None|NotSpecified n_out:
     :rtype: Data
     """
-    if out_type or n_out is not NotSpecified:
+    if out_type or (n_out is not NotSpecified and sources and n_out != sources[0].output.dim):
```

Comment:
This is about `CopyLayer`. Why this change?
(Maybe actually this whole `if` branch could be removed; it's a bit historical. But anyway, it seems unrelated to the remaining PR.)

Reply:
Sorry, I actually made some comments on Friday, but somehow they weren't posted.

I ran into the problem that `n_out` is set via `transform_config_dict` if the layer has a target. Using `super(CopyLayer, cls).get_out_data_from_opts` in this case fails to set `sparse=True`. (This happened copying from a layer in the rec unit to outside, so it's somewhat related.)

With this, I only use the base method if `n_out` is "unexpected", i.e. different than what get_concat_sources_data_template would result in.

I would be happy to remove the `if`. Because in a CopyLayer we should always know the shape from the inputs, so I don't see when setting `out_type` should be necessary.

Reply:
Anyway. This seems unrelated to this pull request. So don't mix it in here.

But maybe create a new pull request, where you remove the whole `if` branch, and then we can see if any test case breaks.